File: backend/src/main/java/com/travelcompanion/service/WeatherService.java
package com.travelcompanion.service;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.util.Map;

@Service
public class WeatherService {

    @Value("${openweather.api.key}")
    private String apiKey;

    @Value("${openweather.api.url}")
    private String apiUrl;

    private final RestTemplate restTemplate = new RestTemplate();

    @SuppressWarnings("unchecked")
    public Map<String, Object> getWeather(String city) {
        String url = apiUrl + "/weather?q=" + city + "&appid=" + apiKey + "&units=metric";
        try {
            return restTemplate.getForObject(url, Map.class);
        } catch (Exception e) {
            // Return mock data if API key is not configured
            return Map.of(
                "name", city,
                "main", Map.of("temp", 25.0, "humidity", 60, "feels_like", 27.0),
                "weather", java.util.List.of(Map.of("main", "Clear", "description", "clear sky", "icon", "01d")),
                "wind", Map.of("speed", 3.5),
                "mock", true
            );
        }
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> getForecast(String city) {
        String url = apiUrl + "/forecast?q=" + city + "&appid=" + apiKey + "&units=metric&cnt=5";
        try {
            return restTemplate.getForObject(url, Map.class);
        } catch (Exception e) {
            return Map.of("city", Map.of("name", city), "list", java.util.List.of(), "mock", true);
        }
    }
}

File: backend/src/services/weather.service.js
const fetch = require('node-fetch');
const config = require('../config');

exports.getWeather = async (city) => {
  if (!config.openweather.apiKey) {
    return getMockWeather(city);
  }

  try {
    const url = `${config.openweather.apiUrl}/weather?q=${encodeURIComponent(city)}&appid=${config.openweather.apiKey}&units=metric`;
    const response = await fetch(url);
    if (!response.ok) {
      return getMockWeather(city);
    }
    return await response.json();
  } catch (error) {
    return getMockWeather(city);
  }
};

exports.getForecast = async (city) => {
  if (!config.openweather.apiKey) {
    return getMockForecast(city);
  }

  try {
    const url = `${config.openweather.apiUrl}/forecast?q=${encodeURIComponent(city)}&appid=${config.openweather.apiKey}&units=metric`;
    const response = await fetch(url);
    if (!response.ok) {
      return getMockForecast(city);
    }
    return await response.json();
  } catch (error) {
    return getMockForecast(city);
  }
};

function getMockWeather(city) {
  return {
    name: city,
    main: { temp: 25, humidity: 60, feels_like: 27 },
    weather: [{ main: 'Clear', description: 'clear sky', icon: '01d' }],
    wind: { speed: 3.5 },
    mock: true
  };
}

function getMockForecast(city) {
  return {
    city: { name: city },
    list: [
      {
        dt_txt: new Date(Date.now() + 86400000).toISOString(),
        main: { temp: 26 },
        weather: [{ description: 'few clouds', icon: '02d' }]
      },
      {
        dt_txt: new Date(Date.now() + 172800000).toISOString(),
        main: { temp: 24 },
        weather: [{ description: 'scattered clouds', icon: '03d' }]
      }
    ],
    mock: true
  };
}
